### PR TITLE
Support AIX in Rust standard library

### PR DESCRIPTION
**Pull request by ecnelises**
Also containing original contributions from @bzEq .

### CODE REVIEW

File: library/std/src/os/aix/process.rs
[the remark this answers is not in the document]
Comment (ecnelises) [reply library/std/src/os/aix/process.rs:1]: Thanks, updated. AIX indeed has some compatibility for Linux specific API, but not including pidfd.

**Comment by ecnelises** [comment]
Gentle ping for more comments...?

**Comment by workingjubilee** [comment]
Is there a reason the powerpc64-ibm-aix target does not have a platform-support.md doc which identifies its maintainers and other appropriate notes about compiling for the target as of this PR?

**Comment by ecnelises** [comment]
> Is there a reason the powerpc64-ibm-aix target does not have a platform-support.md doc which identifies its maintainers and other appropriate notes about compiling for the target as of this PR?

Thanks, created https://github.com/rust-lang/rust/pull/114316. When this target was initially added, I thought there are not such necessary things to mention, especially when cross-compilation is not available yet.

**Comment by workingjubilee** [comment]
Its completeness isn't too important from my POV, as a few targets landed a mere stub for such docs at first and then when they actually had full-ish support the "real" docs were added, I just wanted to make sure it was going to get sorted.

**Comment by kmicklas** [comment]
As an outside data point, I tried this with a few basic programs and everything seemed to be working fine!

Are there any remaining blockers to getting this merged?

File: library/std/src/sys/unix/env.rs [hunk 1]
```diff
@@ -261,3 +261,14 @@ pub mod os {
     pub const EXE_SUFFIX: &str = "";
     pub const EXE_EXTENSION: &str = "";
 }
+
+#[cfg(target_os = "aix")]
+pub mod os {
+    pub const FAMILY: &str = "unix";
+    pub const OS: &str = "aix";
+    pub const DLL_PREFIX: &str = "lib";
+    pub const DLL_SUFFIX: &str = ".a";
```
Comment (workingjubilee) [inline library/std/src/sys/unix/env.rs:270]: **Really**?
Interesting.

File: library/std/src/sys/unix/os.rs [hunk 2]
```diff
@@ -254,6 +255,41 @@ impl StdError for JoinPathsError {
...
+    if let Some(p) = getenv(OsStr::from_bytes("PATH".as_bytes())) {
+        for search_path in split_paths(&p) {
+            let pb = search_path.join(&path);
+            if pb.is_file() && let Ok(metadata) = crate::fs::metadata(&pb) &&
+               metadata.permissions().mode() & 0o111 != 0 {
+                return pb.canonicalize();
+            }
+        }
+    }
+    return Err(io::const_io_error!(ErrorKind::Uncategorized, "an executable path was not found"));
```
Comment (workingjubilee) [inline library/std/src/sys/unix/os.rs:290]: Isn't this the classic case for `ErrorKind::NotFound`? I mean, I guess we could not. I don't think anyone has given the variant used any thought before, so I'd appreciate your feedback, either way.

```suggestion
    Err(io::const_io_error!(ErrorKind::NotFound, "an executable path was not found"))
```

File: library/std/src/sys/unix/os.rs [hunk 3]
```diff
@@ -254,6 +255,41 @@ impl StdError for JoinPathsError {
...
+    #[cfg(test)]
+    use realstd::env;
+
+    #[cfg(not(test))]
+    use crate::env;
+
+    let exe_path = env::args().next().ok_or(io::const_io_error!(
+        ErrorKind::Uncategorized,
+        "an executable path was not found because no arguments were provided through argv"
+    ))?;
```
Comment (workingjubilee) [inline library/std/src/sys/unix/os.rs:271]: Same "isn't this NotFound, kinda?" remark, but I'm happier to leave this one.

**Review by workingjubilee — COMMENTED** [review]
Unix target but with its own folder? Sure, why not.

This looks good to me with or without the mentioned change, but I'd like to hear your thoughts first before proceeding.

**Comment by workingjubilee** [comment]
r? workingjubilee

**Comment by workingjubilee** [comment]
r=me with or without the NotFound change, because Uncategorized errors explicitly disclaim that they can be changed later.
@rustbot author
@bors delegate+

File: library/std/src/sys/unix/env.rs (hunk 1 above)
Comment (ecnelises) [reply library/std/src/sys/unix/env.rs:270]: Looks a little weird but true. `.so` also exists but by convention many dynamic libraries contain `.so` inside `.a`.

**Comment by workingjubilee** [comment]
huh.

**Comment by workingjubilee** [comment]
A completely unrelated PR also failed on this point so
@bors retry

**Comment by ecnelises** [comment]
Thanks!

**Comment by workingjubilee** [comment]
Thank _you_!

**Pull request merged**